src/libstore/derived-path.cc

```diff
@@ -93,24 +93,25 @@ DerivedPath::Opaque DerivedPath::Opaque::parse(const Store & store, std::string_
     return {store.parseStorePath(s)};
 }
 
-DerivedPath::Built DerivedPath::Built::parse(const Store & store, std::string_view s)
+DerivedPath::Built DerivedPath::Built::parse(const Store & store, std::string_view drvS, std::string_view outputsS)
```

Comment:
See https://github.com/NixOS/nix/pull/6815 which I made so this deduplication can happen. Note that part of the complication is `!` is still used in the remote store protocol for `DerivedPath`, so some trickiness is needed so `^` and `!` are used for different use-cases, but we can do better de-duplicating than we do currently.

Reply:
@edolstra can we return to this after this PR? #6815 to deduplicate things has prooved to be much bigger than this. I would really appreciate being able to make this change and then do the clean up afterwards.

Reply:
A simpler patch is not available because we still need to parse with with both `!` and `^`, because the former is still used internally e.g. in the protocols.

Reply:
https://github.com/NixOS/nix/pull/6815 now works! But I'll still keep it separately because that is much bigger than this.